services/requester/batch_tx_pool.go

```diff
@@ -0,0 +1,286 @@
+package requester
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/onflow/cadence"
+	"github.com/onflow/flow-go-sdk"
+	flowGo "github.com/onflow/flow-go/model/flow"
+	gethCommon "github.com/onflow/go-ethereum/common"
+	gethTypes "github.com/onflow/go-ethereum/core/types"
+	"github.com/rs/zerolog"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/onflow/flow-evm-gateway/config"
+	"github.com/onflow/flow-evm-gateway/metrics"
+	"github.com/onflow/flow-evm-gateway/models"
+	"github.com/onflow/flow-evm-gateway/services/requester/keystore"
+)
+
+type pooledEvmTx struct {
+	txPayload cadence.String
+	nonce     uint64
+}
+
+// BatchTxPool is a `TxPool` implementation that collects and groups
+// transactions based on their EOA signer, and submits them for execution
+// using a batch.
+//
+// The underlying Cadence EVM API used, is `EVM.batchRun`, instead of the
+// `EVM.run` used in `SingleTxPool`.
+//
+// The main advantage of this implementation over the `SingleTxPool`, is the
+// guarantee that transactions originated from the same EOA address, which
+// arrive in a short time interval (about the same as Flow's block productionrate),
+// will be executed in the same order their arrived.
+// This helps to reduce the nonce mismatch errors which mainly occur from the
+// re-ordering of Cadence transactions that happens from Collection nodes.
+type BatchTxPool struct {
+	*SingleTxPool
+	pooledTxs map[gethCommon.Address][]pooledEvmTx
+	txMux     sync.Mutex
+	ctx       context.Context
```

Comment:
it's a best practice not to store a context within a struct. instead pass it into the function who's lifecycle will be managed.

Reply:
Good point :+1: Updated in https://github.com/onflow/flow-evm-gateway/pull/831/commits/2a98db83c0099a71511819ff0d75d3747fde5700 .